engines/e_pkcs11.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/engine.h>
+#include <openssl/err.h>
+#include <openssl/crypto.h>
+#include <openssl/rsa.h>
+#include <openssl/bn.h>
+#include <openssl/bio.h>
+#include <internal/dso.h>
+#include <internal/nelem.h>
+
+#define CK_PTR *
+
+#ifdef _WIN32
+# pragma pack(push, cryptoki, 1)
+# define CK_DECLARE_FUNCTION(returnType, name) \
+   returnType __declspec(dllimport) name
+# define CK_DECLARE_FUNCTION_POINTER(returnType, name) \
+   returnType __declspec(dllimport) (* name)
+#else
+# define CK_DECLARE_FUNCTION(returnType, name) \
+   returnType name
+# define CK_DECLARE_FUNCTION_POINTER(returnType, name) \
+   returnType (* name)
+#endif
+
+#define CK_CALLBACK_FUNCTION(returnType, name) \
+   returnType (* name)
+
+#ifndef NULL_PTR
+# define NULL_PTR 0
+#endif
+
+#include "pkcs11.h"
+
+#ifdef _WIN32
+# pragma pack(pop, cryptoki)
+#endif
+
+#define PKCS11_CMD_MODULE_PATH            ENGINE_CMD_BASE
+#define PKCS11_CMD_PIN                    (ENGINE_CMD_BASE + 1)
+
+static const ENGINE_CMD_DEFN pkcs11_cmd_defns[] = {
+    {PKCS11_CMD_MODULE_PATH,
+     "MODULE_PATH",
+     "Module path",
+     ENGINE_CMD_FLAG_STRING},
+    {PKCS11_CMD_PIN,
+     "PIN",
+     "PIN",
+     ENGINE_CMD_FLAG_STRING},
+    {0, NULL, NULL, 0}
+};
+
+struct X509_sig_st {
```

Comment:
Oups:  this is a workaround over "Opaque Structures" feature.
You better use official API to build  a X509_SIG instance.

Reply:
I think it is useful for others to make public the internal method encode_pkcs1 which is located inside crypto / rsa / rsa_sign.c.
If you want I can send a new PR only for this change (internal) encode_pkcs1 to (public) RSA_encode_pkcs1